packages/@ember/-internals/runtime/lib/system/core_object.js

```diff
@@ -135,6 +135,49 @@ function initialize(obj, properties) {
 }
 
 /**
+  `CoreObject` is the base class for all Ember constructs. It establishes a
+  class system based on Ember's Mixin system, and provides the basis for the
+  Ember Object Model. `CoreObject` should generally not be used directly,
+  instead you should use `EmberObject`.
+
+  ## Usage
+
+  You can define a class by extending from `CoreObject` using the `extend`
+  method:
+
+  ```js
+  const Person = CoreObject.extend({
+    name: 'Tomster',
+  });
+  ```
+
+  For detailed usage, see the [Object Model](https://guides.emberjs.com/release/object-model/)
+  section of the guides.
+
+  ## Usage with Native Classes
+
+  Native Javascript `class` syntax can be used to extend from any `CoreObject`
+  based class:
+
+  ```js
+  class Person extends CoreObject {
+    init() {
+      super.init(...arguments);
+      this.name = 'Tomster';
+    }
+  }
+  ```
+
+  Some notes about `class` usage:
+
+  * Instead of using `this._super()`, you should use standard `super` syntax in
+    native classes. See the [MDN docs on classes](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Classes#Super_class_calls_with_super)
+    for more details.
+  * Avoid using the `constructor` in favor of `init`. Services are not available
+    during the `constructor` phase, and the nuances of the two can be confusing.
+  * Using native classes, and switching back to the old Ember Object model is
+    fully supported.
```

Comment:
Is this supported/encouraged/discouraged?

```ts
export default class XFoo extends Ember.Component({

  // properties that I deliberately want to share between instances
  tagName: 'ul',

  fullName: computed('firstName', 'lastName', function() {
	return `${this.firstName} ${this.lastName}`;
  }),

  // immutable "defaults" for member data, deliberately shared across all instances
  firstName: '',  
  lastName: ''

}) {
 
  // standard ES6 class stuff

  logName() {
    console.log(this.fullName);
  }
}

```

Reply:
Discouraged, since it will be very difficult to codemod half-way states like this. It's much easier for us to go from a clean-slate to full ES classes in the codemod.

Reply:
I don't think this should be addressed in here, but in a linting rule. We shouldn't even suggest this is a possibility in our docs, ideally. We could cover it in guides though.